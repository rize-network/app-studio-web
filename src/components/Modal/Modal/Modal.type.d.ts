export type Position = 'top' | 'left' | 'right' | 'bottom' | 'center';

export type Shape = 'sharp' | 'rounded';

export type CloseButtonPosition = 'left' | 'right' | 'none';

export type Size = 'xs' | 'sm' | 'md' | 'lg' | 'xl';
