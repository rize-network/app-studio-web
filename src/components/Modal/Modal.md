### **Import**

```tsx static
import { Modal } from 'app-studio';
```

### **Default**

It uses the isOpen and onClose properties to display the modal.

```tsx
import { Button } from '../Button/Button';
import { useState } from 'react';

const [show, setShow] = useState(false);

<>
  <Button onClick={() => setShow(true)} isAuto>
    Open Modal
  </Button>
  <Modal.Overlay isOpen={show} onClose={() => setShow(false)}>
    <Modal.Container>
      <Modal.Header>Title</Modal.Header>
      <Modal.Body>
        Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam.
        Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque
        nisl consectetur et.
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={() => setShow(false)}>Cancel</Button>
      </Modal.Footer>
    </Modal.Container>
  </Modal.Overlay>
</>;
```

## **Modal Overlay**

It represents the overlay or backdrop that appears behind the modal content when the modal is displayed.

### **FullScreen**

“**is*fullScreen***” property changes the width and height of the modal so that it fits the entire screen.

```tsx
import { Button } from '../Button/Button';
import { useState } from 'react';

const [show, setShow] = useState(false);

<>
  <Button onClick={() => setShow(true)} isAuto>
    Full Screen
  </Button>
  <Modal.Overlay isOpen={show} onClose={() => setShow(false)}>
    <Modal.Container isFullScreen>
      <Modal.Header>Title</Modal.Header>
      <Modal.Body>
        Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam.
        Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque
        nisl consectetur et.
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={() => setShow(false)}>Cancel</Button>
      </Modal.Footer>
    </Modal.Container>
  </Modal.Overlay>
</>;
```

### **Position**

“**_position_**” places the modal in the specified direction. It has a default positioning of “center”.

```tsx
import { useState } from 'react';
import { Button } from '../Button/Button';
import { Horizontal } from '../Layout/Horizontal/Horizontal';

const [showTop, setShowTop] = useState(false);
const [showBottom, setShowBottom] = useState(false);
const [showRight, setShowRight] = useState(false);
const [showLeft, setShowLeft] = useState(false);
const [showCenter, setShowCenter] = useState(false);

<>
  <Horizontal gap={10} wrap="nowrap">
    <Button onClick={() => setShowTop(true)}>Top</Button>
    <Button onClick={() => setShowBottom(true)}>Bottom</Button>
    <Button onClick={() => setShowRight(true)}>Right</Button>
    <Button onClick={() => setShowLeft(true)}>Left</Button>
    <Button onClick={() => setShowCenter(true)}>Center</Button>
  </Horizontal>

  {showBottom && (
    <Modal.Overlay isOpen={showBottom} onClose={() => setShowBottom(false)} position="bottom">
      <Modal.Container>
        <Modal.Header>Title</Modal.Header>
        <Modal.Body>
          Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam.
          Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque
          nisl consectetur et.
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={() => setShowBottom(false)}>Cancel</Button>
        </Modal.Footer>
      </Modal.Container>
    </Modal.Overlay>
  )}
  {showTop && (
    <Modal.Overlay isOpen={showTop} onClose={() => setShowTop(false)} position="top">
      <Modal.Container>
        <Modal.Header>Title</Modal.Header>
        <Modal.Body>
          Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam.
          Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque
          nisl consectetur et.
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={() => setShowTop(false)}>Cancel</Button>
        </Modal.Footer>
      </Modal.Container>
    </Modal.Overlay>
  )}
  {showLeft && (
    <Modal.Overlay isOpen={showLeft} onClose={() => setShowLeft(false)} position="left">
      <Modal.Container>
        <Modal.Header>Title</Modal.Header>
        <Modal.Body>
          Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam.
          Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque
          nisl consectetur et.
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={() => setShowLeft(false)}>Cancel</Button>
        </Modal.Footer>
      </Modal.Container>
    </Modal.Overlay>
  )}
  {showRight && (
    <Modal.Overlay isOpen={showRight} onClose={() => setShowRight(false)} position="right">
      <Modal.Container>
        <Modal.Header>Title</Modal.Header>
        <Modal.Body>
          Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam.
          Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque
          nisl consectetur et.
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={() => setShowRight(false)}>Cancel</Button>
        </Modal.Footer>
      </Modal.Container>
    </Modal.Overlay>
  )}
  {showCenter && (
    <Modal.Overlay isOpen={showCenter} onClose={() => setShowCenter(false)} position="center">
      <Modal.Container>
        <Modal.Header>Title</Modal.Header>
        <Modal.Body>
          Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam.
          Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque
          nisl consectetur et.
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={() => setShowCenter(false)}>Cancel</Button>
        </Modal.Footer>
      </Modal.Container>
    </Modal.Overlay>
  )}
</>;
```

### **Blur**

“**_blur_**” adds a blur effect on the modal overlay.

```tsx
import { useState } from 'react';
import { Button } from '../Button/Button';

const [show, setShow] = useState(false);

<>
  <Button onClick={() => setShow(true)}>Blur Overlay</Button>
  <Modal.Overlay isOpen={show} onClose={() => setShow(false)} blur={10}>
    <Modal.Container>
      <Modal.Header>Title</Modal.Header>
      <Modal.Body>
        Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam.
        Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque
        nisl consectetur et.
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={() => setShow(false)}>Cancel</Button>
      </Modal.Footer>
    </Modal.Container>
  </Modal.Overlay>
</>;
```

### **Prevent Close**

“**_isClosePrevented_**” prevents the user from closing the modal when clicking outside of it.

```tsx
import { useState } from 'react';
import { Button } from '../Button/Button';

const [show, setShow] = useState(false);

<>
  <Button onClick={() => setShow(true)} isAuto>
    Prevent Close on Overlay
  </Button>
  <Modal.Overlay isOpen={show} onClose={() => setShow(false)} isClosePrevented>
    <Modal.Container>
      <Modal.Header>Title</Modal.Header>
      <Modal.Body>
        Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget quam.
        Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel scelerisque
        nisl consectetur et.
      </Modal.Body>
      <Modal.Footer>
        <Button onClick={() => setShow(false)}>Cancel</Button>
      </Modal.Footer>
    </Modal.Container>
  </Modal.Overlay>
</>;
```

## **Modal Container**

It represents the main container or wrapper that holds all the content and components of a modal.

### **Shapes**

“**_shape_**” gives a sharp or rounded edges to the modal container.

```tsx
import { useState } from 'react';
import { Button } from '../Button/Button';
import { Text } from '../Text/Text';
import { Horizontal } from '../Layout/Horizontal/Horizontal';

const [showSharp, setShowSharp] = useState(false);
const [showRounded, setShowRounded] = useState(false);

<>
  <Horizontal gap={10}>
    <Button onClick={() => setShowSharp(true)}>Sharp</Button>
    <Button onClick={() => setShowRounded(true)}>Rounded</Button>
  </Horizontal>

  {showSharp && (
    <Modal.Overlay isOpen={showSharp} onClose={() => setShowSharp(false)}>
      <Modal.Container variant="sharp">
        <Modal.Body>
          <Text isTruncated maxLines={2}>
            Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget
            quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel
            scelerisque nisl consectetur et.
          </Text>
        </Modal.Body>
      </Modal.Container>
    </Modal.Overlay>
  )}
  {showRounded && (
    <Modal.Overlay isOpen={showRounded} onClose={() => setShowRounded(false)}>
      <Modal.Container variant="rounded">
        <Modal.Body>
          <Text isTruncated maxLines={2}>
            Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget
            quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel
            scelerisque nisl consectetur et.
          </Text>
        </Modal.Body>
      </Modal.Container>
    </Modal.Overlay>
  )}
</>;
```

### **Shadow**

“**_shadow_**” adds a shadow effect on the modal container.

```tsx
import { useState } from 'react';
import { Button } from '../Button/Button';
import { Text } from '../Text/Text';

const [show, setShow] = useState(false);

<>
  <Button onClick={() => setShow(true)}>Shadow</Button>

  {show && (
    <Modal.Overlay isOpen={show} onClose={() => setShow(false)}>
      <Modal.Container shadow={{ boxShadow: '0px 2px 8px rgba(0, 0, 0, 0.6)' }}>
        <Modal.Body>
          <Text isTruncated maxLines={2}>
            Cras mattis consectetur purus sit amet fermentum. Cras justo odio, dapibus ac facilisis in, egestas eget
            quam. Morbi leo risus, porta ac consectetur ac, vestibulum at eros. Praesent commodo cursus magna, vel
            scelerisque nisl consectetur et.
          </Text>
        </Modal.Body>
      </Modal.Container>
    </Modal.Overlay>
  )}
</>;
```

## **Modal Header**

It represents the header section of a modal, which typically contains a title or heading and optionally other elements such as a close button or additional controls.

### **Close Button Alignment**

“**_buttonPosition_**” places the closing button found in the header modal in the specified direction. It has a default position of “**right**”.

```tsx
import { useState } from 'react';
import { Button } from '../Button/Button';
import { Horizontal } from '../Layout/Horizontal/Horizontal';

const [showRight, setShowRight] = useState(false);
const [showWithout, setShowWithout] = useState(false);
const [showLeft, setShowLeft] = useState(false);

<>
  <Horizontal gap={10}>
    <Button onClick={() => setShowRight(true)} isAuto>
      Close button on the right
    </Button>
    <Button onClick={() => setShowWithout(true)} isAuto>
      Without Close Button
    </Button>
    <Button onClick={() => setShowLeft(true)} isAuto>
      Without Close left
    </Button>
  </Horizontal>
  {showRight && (
    <Modal.Overlay isOpen={showRight} onClose={() => setShowRight(false)}>
      <Modal.Container>
        <Modal.Header> Close Button</Modal.Header>
        <Modal.Body>Cras mattis consectetur purus sit amet fermentum.</Modal.Body>
        <Modal.Footer>
          <Button onClick={() => setShowRight(false)}>Cancel</Button>
        </Modal.Footer>
      </Modal.Container>
    </Modal.Overlay>
  )}
  {showWithout && (
    <Modal.Overlay isOpen={showWithout} onClose={() => setShowWithout(false)}>
      <Modal.Container>
        <Modal.Header buttonPosition="none"> Close Button</Modal.Header>
        <Modal.Body>Cras mattis consectetur purus sit amet fermentum.</Modal.Body>
        <Modal.Footer>
          <Button onClick={() => setShowWithout(false)}>Cancel</Button>
        </Modal.Footer>
      </Modal.Container>
    </Modal.Overlay>
  )}
  {showLeft && (
    <Modal.Overlay isOpen={showLeft} onClose={() => setShowLeft(false)}>
      <Modal.Container>
        <Modal.Header buttonPosition="left">Close Button</Modal.Header>
        <Modal.Body>Cras mattis consectetur purus sit amet fermentum.</Modal.Body>
        <Modal.Footer>
          <Button onClick={() => setShowLeft(false)}>Cancel</Button>
        </Modal.Footer>
      </Modal.Container>
    </Modal.Overlay>
  )}
</>;
```

## **Modal Body**

It represents the main content area of a modal, where the primary information or user interaction elements are displayed.

### **LargeContent**

A vertical scroll is displayed, if the content height exceed the body height.

```tsx
import { useState } from 'react';
import { Button } from '../Button/Button';
import { Text } from '../Text/Text';

const [show, setShow] = useState(false);

<>
  <Button onClick={() => setShow(true)}>Scroll</Button>
  {show && (
    <Modal.Overlay isOpen={show} onClose={() => setShow(false)}>
      <Modal.Container>
        <Modal.Header buttonPosition="none">
          <Text size="lg" weight="semibold">
            Title
          </Text>
        </Modal.Header>
        <Modal.Body height="200px">
          Cras mattis consectetur purus sit amet fermentum.Cras mattis consectetur purus sit amet fermentum.Cras mattis
          consectetur purus sit amet fermentum.Cras mattis consectetur purus sit amet fermentum.Cras mattis consectetur
          purus sit amet fermentum.Cras mattis consectetur purus sit amet fermentum.Cras mattis consectetur purus sit
          amet fermentum.Cras mattis consectetur purus sit amet fermentum.Cras mattis consectetur purus sit amet Cras
          mattis consectetur purus sit amet fermentum.Cras mattis consectetur purus sit amet fermentum.Cras mattis
          consectetur purus sit amet fermentum.Cras mattis consectetur purus sit amet fermentum.Cras mattis consectetur
          purus sit amet fermentum.Cras mattis consectetur purus sit amet fermentum.Cras mattis consectetur purus sit
          amet fermentum.Cras mattis consectetur purus sit amet fermentum.Cras mattis consectetur purus sit amet
          fermentum.Cras mattis consectetur purus sit amet fermentum.
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={() => setShow(false)}>Cancel</Button>
        </Modal.Footer>
      </Modal.Container>
    </Modal.Overlay>
  )}
</>;
```

## **Modal Footer**

It represents the footer section of a modal, which typically contains buttons or controls for actions related to the modal's content.

```tsx
import { useState } from 'react';
import { Button } from '../Button/Button';
import { Text } from '../Text/Text';

const [show, setShow] = useState(false);

<>
  <Button onClick={() => setShow(true)}>Footer</Button>
  {show && (
    <Modal.Overlay isOpen={show} onClose={() => setShow(false)}>
      <Modal.Container>
        <Modal.Header buttonPosition="none">
          <Text size="lg" weight="semibold">
            Title
          </Text>
        </Modal.Header>
        <Modal.Body height="200px">
          Cras mattis consectetur purus sit amet fermentum.Cras mattis consectetur purus sit amet fermentum.Cras mattis
          consectetur purus sit amet fermentum.Cras mattis consectetur purus sit amet fermentum.Cras mattis consectetur
          purus sit amet fermentum.
        </Modal.Body>
        <Modal.Footer>
          <Button onClick={() => setShow(false)}>Cancel</Button>
        </Modal.Footer>
      </Modal.Container>
    </Modal.Overlay>
  )}
</>;
```

## **Types**

```tsx static
type Position = 'top' | 'left' | 'right' | 'bottom' | 'center';
```

```tsx static
type Shape = 'sharp' | 'rounded';
```

```tsx static
type CloseButtonPosition = 'left' | 'right' | 'none';
```

```tsx static
type Size = 'xs' | 'sm' | 'md' | 'lg' | 'xl';
```

## **Default Styles**

#### **OverlayAlignments**

```tsx static
const OverlayAlignments: Record<Position, CSSProperties> = {
  center: { justifyContent: 'center', alignItems: 'center' },
  top: { justifyContent: 'center' },
  right: { justifyContent: 'flex-end', alignItems: 'center' },
  bottom: { justifyContent: 'center', alignItems: 'flex-end' },
  left: { alignItems: 'center' },
};
```

#### **ContainerShapes**

```tsx static
const ContainerShapes: Record<Shape, CSSProperties> = {
  sharp: { borderRadius: 0 },
  rounded: { borderRadius: 4 },
};
```

#### **HeaderIconSizes**

```tsx static
const HeaderIconSizes: Record<Size, number> = {
  xs: 12,
  sm: 16,
  md: 20,
  lg: 24,
  xl: 28,
};
```
