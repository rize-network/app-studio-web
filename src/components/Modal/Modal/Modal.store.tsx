import { create } from 'zustand';

export interface ModalState {
  modal: string | boolean;
  modalProps: any & { isVisible: boolean };
  overlayProps: any;
  show: (modal: string, modalProps?: any, overlayProps?: any) => void;
  hide: () => void;
}

export const useModalStore = create<ModalState>((set) => ({
  modal: false,
  modalProps: { isVisible: false },
  overlayProps: {},
  show: (modal, modalProps = {}, overlayProps = {}) => {
    if (modal) {
      modalProps.isVisible = true;
    }
    set((state: ModalState) => ({
      ...state,
      modal,
      modalProps,
      overlayProps,
    }));
  },
  hide: () => {
    set((state: ModalState) => ({
      ...state,
      modalProps: {
        isVisible: false,
      },
    }));
  },
}));

export const showModal = (
  modal: string,
  modalProps: any = {},
  overlayProps: any = {}
) => {
  useModalStore.getState().show(modal, modalProps, overlayProps);
};

export const hideModal = () => {
  useModalStore.getState().hide();
};
