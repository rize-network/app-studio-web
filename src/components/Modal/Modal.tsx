import React from 'react';

import { ModalType } from './Modal/Modal.props';
import {
  ModalBody,
  ModalContainer,
  ModalFooter,
  ModalHeader,
  ModalOverlay,
} from './Modal/Modal.view';
import { ModalLayout } from './Modal/Modal.layout';

/**
 * It is a custom content overlay that appears on top of the main screen.
 */
// eslint-disable-next-line react/prop-types
export const Modal: ModalType = ({ children }) => <>{children}</>;

Modal.Overlay = ModalOverlay;
Modal.Container = ModalContainer;
Modal.Header = ModalHeader;
Modal.Body = ModalBody;
Modal.Footer = ModalFooter;
Modal.Layout = ModalLayout;

export * from './Modal/Modal.store';
