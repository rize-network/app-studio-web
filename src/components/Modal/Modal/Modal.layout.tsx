import React, { Fragment } from 'react';

import { ModalState, hideModal, useModalStore } from './Modal.store';
import { ModalOverlay } from './Modal.view';
import { ModalLayoutProps } from './Modal.props';

export const ModalLayout = ({ modals }: ModalLayoutProps) => {
  const modalStore = useModalStore(
    ({ modal, modalProps, overlayProps }: ModalState) => ({
      modal,
      modalProps,
      overlayProps,
    })
  );
  if (typeof modalStore.modal === 'boolean') {
    return null;
  }

  const Modal = modals[modalStore.modal];

  if (!Modal) {
    console.error(modalStore.modal + " modal doesn't exist");
    return null;
  }

  return (
    <Fragment>
      <ModalOverlay
        isOpen={modalStore.modalProps.isVisible}
        onClose={hideModal}
        blur={10}
        {...modalStore.overlayProps}
      >
        { Modal !== undefined && <Modal {...modalStore.modalProps} />}
      </ModalOverlay>
    </Fragment>
  );
};
