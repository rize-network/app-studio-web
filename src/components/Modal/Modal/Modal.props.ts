import React from 'react';
import { Elevation } from '../../../utils/elevation';
import { Shadow, ViewProps } from 'app-studio';

import { CloseButtonPosition, Position, Shape, Size } from './Modal.type';

export interface ModalProps extends Omit<ViewProps, 'size'> {
  /**
   * The color of the close button in the header.
   */
  buttonColor?: string;

  /**
   * The size of the close button in the header.
   */
  iconSize?: Size;

  /**
   * Determines whether the modal should have sharp or rounded edges.
   */
  shape?: Shape;

  /**
   * If set to true, the modal will occupy the full width and height of the screen.
   */
  isFullScreen?: boolean;

  /**
   * Indicates the presence of a close button and its position within the header.
   */
  buttonPosition?: CloseButtonPosition;

  /**
   * The callback function to be executed when the close button is clicked or pressed.
   */
  onClose?: () => void;

  /**
   * The degree of blurriness applied to the overlay.
   */
  blur?: number;

  /**
   * If set to true, the modal will be visible.
   */
  isOpen?: boolean;

  /**
   * If set to true, the modal will remain open when the overlay is clicked.
   */
  isClosePrevented?: boolean;

  /**
   * The positioning of the modal container.
   */
  position?: Position;

  /**
   * Applies a shadow effect to the button.
   */
  shadow?: Shadow | Elevation | ViewProps;
}

export interface ModalType extends React.FunctionComponent<ModalProps> {
  /**
   * The overlay component of the modal.
   */
  Overlay: React.FC<OverlayProps>;

  /**
   * The container component of the modal.
   */
  Container: React.FC<ContainerProps>;

  /**
   * The header component of the modal.
   */
  Header: React.FC<HeaderProps>;

  /**
   * The body component of the modal.
   */
  Body: React.FC<BodyProps>;

  /**
   * The footer component of the modal.
   */
  Footer: React.FC<FooterProps>;

  /**
   * The header component of the modal.
   */
  Layout: React.FC<ModalLayoutProps>;
}

export interface OverlayProps extends Omit<ViewProps, 'position' | 'isOpen'> {
  /**
   * The content of the Overlay
   */
  children?: React.ReactNode;
  /**
   * The value of how much to blur the overlay
   */
  blur?: number;
  /**
   * If true, the modal will be visible
   */
  isOpen: boolean;
  /**
   * If true, the modal will not closed when the overlay is clicked
   */
  isClosePrevented?: boolean;
  /**
   * To position the modal container
   */
  position?: Position;
  /**
   * Action to be handled when the button close is clicked or pressed
   */
  onClose: () => void;
}

export interface HeaderProps extends Omit<ViewProps, 'size'> {
  views?: {
    container?: ViewProps;
    header?: ViewProps;
  };
  /**
   * The content of the header
   */
  buttonColor?: string;
  /**
   * To set the close button size
   */
  iconSize?: Size;
  /**
   * The content of the header
   */
  children?: React.ReactNode;
  /**
   * Indicates if there is a button and where to position it
   */
  buttonPosition?: CloseButtonPosition;
  /**
   * Action to be handled when the button close is clicked or pressed
   */
  onClose?: () => void;
}

export interface ModalLayoutProps {
  modals: { [x: string]: React.FC<any> };
  onShow?: (name: string, props?: any) => void;
  onHide?: (name?: string, props?: any) => void;
}

export interface ContainerProps extends Omit<ViewProps, 'size'> {
  views?: {
    container?: ViewProps;
  };
  /**
   * The content of the modal container
   */
  children: React.ReactNode;
  /**
   * To give a sharp or rounded edges to the modal
   */
  shape?: Shape;
  /**
   * If true, the modal will take the full width and height
   */
  isFullScreen?: boolean;
  /**
   * Set a shadow effect on the button.
   */
  shadow?: Shadow | Elevation | ViewProps;
}

export interface BodyProps extends ViewProps {
  views?: {
    view?: ViewProps;

    /**
     * Optional theme mode override ('light' or 'dark')
     * If not provided, the component will use the theme mode from context
     */
  };
}

export interface FooterProps extends ViewProps {
  views?: {
    container?: ViewProps;
  };
}
