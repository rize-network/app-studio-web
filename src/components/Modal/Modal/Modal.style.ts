/**
 * Modal Styles
 *
 * Defines the styles for the Modal component following the design guidelines:
 * - Typography: Inter/Geist font, specific sizes/weights
 * - Spacing: 4px grid system
 * - Colors: Neutral palette with semantic colors
 * - Rounded corners: Consistent border radius
 * - Transitions: Subtle animations
 */

import { ViewProps } from 'app-studio';
import { Shape, Size } from './Modal.type';

/**
 * Container shape configurations for the Modal component
 * Following the 4px grid system
 */
export const ContainerShapes: Record<Shape, ViewProps> = {
  sharp: { borderRadius: 0 },
  rounded: { borderRadius: '8px' }, // 2 × 4px grid
};
/**
 * Overlay alignment configurations for the Modal component
 */
export const OverlayAlignments: Record<string, ViewProps> = {
  center: { justifyContent: 'center', alignItems: 'center' },
  top: {
    justifyContent: 'center',
    alignItems: 'flex-start',
    paddingTop: '64px',
  }, // 16 × 4px grid
  right: {
    justifyContent: 'flex-end',
    alignItems: 'center',
    paddingRight: '32px',
  }, // 8 × 4px grid
  bottom: {
    justifyContent: 'center',
    alignItems: 'flex-end',
    paddingBottom: '64px',
  }, // 16 × 4px grid
  left: {
    justifyContent: 'flex-start',
    alignItems: 'center',
    paddingLeft: '32px',
  }, // 8 × 4px grid
};

/**
 * Header icon size configurations for the Modal component
 * Following the 4px grid system
 */
export const HeaderIconSizes: Record<Size, number> = {
  xs: 12, // 3 × 4px grid
  sm: 16, // 4 × 4px grid
  md: 20, // 5 × 4px grid
  lg: 24, // 6 × 4px grid
  xl: 28, // 7 × 4px grid
};

/**
 * Animation configurations for the Modal component
 */
export const ModalAnimations = {
  enter: {
    opacity: [0, 1],
    transform: ['scale(0.95)', 'scale(1)'],
    transition: 'all 0.3s ease-out',
  },
  exit: {
    opacity: [1, 0],
    transform: ['scale(1)', 'scale(0.95)'],
    transition: 'all 0.2s ease-in',
  },
};

/**
 * Typography configurations for the Modal component
 */
export const ModalTypography = {
  fontFamily: 'Inter, -apple-system, BlinkMacSystemFont, sans-serif',
  title: {
    fontSize: '18px', // 4.5 × 4px grid
    fontWeight: '600', // Semi-bold
    lineHeight: '24px', // 6 × 4px grid
    color: 'color.gray.900',
  },
  body: {
    fontSize: '16px', // 4 × 4px grid
    fontWeight: '400', // Regular
    lineHeight: '24px', // 6 × 4px grid
    color: 'color.gray.700',
  },
};
