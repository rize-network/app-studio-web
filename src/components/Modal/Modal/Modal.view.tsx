import React, { useEffect } from 'react';
import { useTheme } from 'app-studio';
import { Center, Vertical } from 'src/components';
import { Horizontal } from 'src/components';
import { Button } from 'src/components';
import { View } from 'src/components/Layout/View/View';
import { CloseSvg } from 'src/components/Svg';
import { useModalStore } from 'src/store/useModalStore';

import { BodyProps, ContainerProps, FooterProps, HeaderProps, OverlayProps } from '../Modal/Modal.props';
import { ContainerShapes, OverlayAlignments } from '../Modal/Modal.style';
import { HeaderIconSizes } from '../Modal/Modal.style';

export const ModalOverlay: React.FC<OverlayProps> = ({
  children,
  blur,
  isOpen = false,
  isClosePrevented = false,
  onClose = () => {},
  position = 'center',
  ...props
}) => {
  const setOpen = useModalStore((state: any) => state.setOpen);
  const setOnClose = useModalStore((state: any) => state.setOnClose);

  useEffect(() => {
    setOnClose(onClose);
    if (isOpen) setOpen();
  });

  if (!isOpen) return null;

  const handleClick = () => {
    if (!isClosePrevented) onClose();
  };

  return (
    <Center
      position="fixed"
      top={0}
      left={0}
      width="100%"
      height="100%"
      zIndex={1000}
      onPress={handleClick}
      visibility={isOpen ? 'visible' : 'hidden'}
    >
      <View
        cursor="pointer"
        position="absolute"
        top={0}
        left={0}
        zIndex={1000}
        width="100vw"
        height="100vh"
        display="flex"
        backgroundColor="blackAlpha.500"
        backdropFilter={blur ? `blur(${blur}px)` : undefined}
        onPress={handleClick}
        {...OverlayAlignments[position]}
        {...props}
      >
        {children}
      </View>
    </Center>
  );
};

export const ModalContainer: React.FC<ContainerProps> = ({
  children,
  shadow,
  isFullScreen = false,
  shape = 'rounded',
  ...props
}) => {
  const defaultShadow =
    typeof document !== undefined
      ? {
          boxShadow: '0px 2px 8px rgba(0, 0, 0, 0.3)',
        }
      : {
          elevation: 5,
          shadowColor: 'rgba(0, 0, 0, 0.3)',
          shadowOffset: { width: 0, height: 2 },
          shadowOpacity: 1,
          shadowRadius: 8,
        };

  const handleClick = (event: any) => event.stopPropagation();
  return (
    <Vertical
      cursor="default"
      overflow="hidden"
      backgroundColor="white"
      width={isFullScreen ? '100%' : 600}
      height={isFullScreen ? '100%' : 'fit-content'}
      onPress={handleClick}
      {...(shadow ? shadow : defaultShadow)}
      {...ContainerShapes[shape]}
      {...props}
    >
      {children}
    </Vertical>
  );
};

export const ModalHeader: React.FC<HeaderProps> = ({
  children,
  buttonColor = 'theme.primary',
  iconSize = 'md',
  buttonPosition = 'right',
  ...props
}) => {
  const onClose = useModalStore((state: any) => state.onClose);

  const buttonIcon = (
    <Button
      onPress={onClose}
      colorScheme="transparent"
      icon={<CloseSvg size={HeaderIconSizes[iconSize]} color={buttonColor} />}
      padding={0}
      margin={0}
      filter="none"
      isAuto
    />
  );

  return (
    <Horizontal
      justifyContent={buttonPosition === 'none' ? 'center' : 'space-between'}
      alignItems="center"
      paddingVertical={15}
      paddingHorizontal={20}
      {...props}
    >
      {buttonPosition === 'left' && buttonIcon}
      {children}
      {buttonPosition === 'right' && buttonIcon}
    </Horizontal>
  );
};

export const ModalBody: React.FC<BodyProps> = ({ children, ...props }) => {
  const defaultBorder = {
    borderBottomWidth: 2,
    borderTopWidth: 2,
    borderColor: 'rgba(250, 250, 250, 1)',
    borderStyle: 'solid',
  };
  return (
    <View overflowY="auto" paddingVertical={15} paddingHorizontal={20} {...defaultBorder} {...props}>
      {children}
    </View>
  );
};

export const ModalFooter: React.FC<FooterProps> = ({ children, ...props }) => {
  return (
    <Horizontal
      marginTop="auto"
      alignItems="center"
      justifyContent="flex-end"
      paddingVertical={15}
      paddingHorizontal={20}
      {...props}
    >
      {children}
    </Horizontal>
  );
};
