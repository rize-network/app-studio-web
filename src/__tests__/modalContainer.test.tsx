import React from 'react';
import renderer from 'react-test-renderer';
import Modal from 'src/components/Modal/Modal';

import { cleanup, render, screen } from '@testing-library/react';

afterEach(() => {
  cleanup();
});

test('ModalContainer renders without errors when children prop is provided', () => {
  render(
    <Modal.Container>
      <div>Modal Content</div>
    </Modal.Container>
  );
});

test('ModalContainer renders children within the container', () => {
  render(
    <Modal.Container>
      <div>Modal Content</div>
    </Modal.Container>
  );
  expect(screen.getByText('Modal Content')).toBeInTheDocument();
});

test('ModalContainer applies correct default styles', () => {
  render(
    <Modal.Container role="container">
      <div>Modal Content</div>
    </Modal.Container>
  );
  const modalContainer = screen.getByRole('container');
  expect(modalContainer).toHaveStyle('overflow: hidden');
  expect(modalContainer).toHaveStyle('background-color: white');
});

test('ModalContainer to be full screen', () => {
  render(
    <Modal.Container role="container" isFullScreen>
      <div>Modal Content</div>
    </Modal.Container>
  );
  const modalContainer = screen.getByText('Modal Content');
  expect(modalContainer).toHaveStyle('width: 600');
  expect(modalContainer).toHaveStyle('height: fit-content');
});

test('ModalContainer applies custom styles using props spread', () => {
  render(
    <Modal.Container role="container" style={{ color: 'red' }}>
      <div>Modal Content</div>
    </Modal.Container>
  );
  const modalContainer = screen.getByRole('container');
  expect(modalContainer).toHaveStyle('color: red');
});

test('Modal Container to match snapshot', () => {
  const tree = renderer
    .create(
      <Modal.Container role="container" shape="rounded" shadow={{ boxShadow: 'none' }} isFullScreen>
        <div>Modal Content</div>
      </Modal.Container>
    )
    .toJSON();
  expect(tree).toMatchSnapshot();
});
