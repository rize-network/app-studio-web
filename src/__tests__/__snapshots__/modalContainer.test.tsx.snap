// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal Container to match snapshot 1`] = `
<div
  className="sc-beqWaB bAMrnX"
  cursor="default"
  display="flex"
  height="100%"
  onClick={[Function]}
  onPress={[Function]}
  overflow="hidden"
  role="container"
  width="100%"
>
  <div>
    Modal Content
  </div>
</div>
`;
